[SPARK-13250][SQL] Update PhysicallRDD to convert to UnsafeRow if using the vectorized scanner.

nongli opened the pull request:
Some parts of the engine rely on UnsafeRow which the vectorized parquet scanner does not want
to produce. This add a conversion in Physical RDD. In the case where codegen is used (and the
scan is the start of the pipeline), there is no requirement to use UnsafeRow. This patch adds
update PhysicallRDD to support codegen, which eliminates the need for the UnsafeRow conversion
in all cases.

The result of these changes for TPCDS-Q19 at the 10gb sf reduces the query time from 9.5 seconds
to 6.5 seconds.

nongli commented:
@davies I think this is a general purpose way we can remove 'copy' from InternalRow. We're essentially delaying the serialization as late in the plan tree as possible.

davies commented:
In the past, we had a planner to insert a ConvertToUnsafe and ConvertToSafe for operators that expect UnsafeRow or SafeRow, that is removed to simplify complexity, maybe we need to bring that back for the parquet reader? see #10511

SparkQA commented:
**[Test build #50996 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50996/consoleFull)** for PR 11141 at commit [`662559c`](https://github.com/apache/spark/commit/662559c1ef24159d6878616993b2dd3caa3de154).

nongli commented:
@davies I don't like that approach. Converting to unsaferow is not really the thing we want to be doing long term. In this example for instance, i think it would be nice if the value in the hashtable was just the unsafe row of the non-join key columns. This is hard to do if we inserted a plan node.

SparkQA commented:
**[Test build #50996 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50996/consoleFull)** for PR 11141 at commit [`662559c`](https://github.com/apache/spark/commit/662559c1ef24159d6878616993b2dd3caa3de154).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50996/
Test FAILed.

davies commented:
@nongli Are you suggesting that we should do this kind of workaround for every operator that requires UnsafeRow (in-memory cache, SortMergeJoin, CartesianProduct, Exchange, etc) ?

nongli commented:
@davies Yes I am. I don't think we're going to add a ton more operators and in all the one you mentioned, we should think hard about serializing to the in memory structure the operator wants rather than just copying. For example, CartesianProduct should probably serialize all the rows on one side contiguously; similarly sort for tungsten pages. In-memory cache should use a columnar version and not need to first go to UnsafeRow.

I think we can take what I've done here and make it more componentized so it's less code duplication to use this elsewhere. I'm okay not doing this in general in the planner. The operators that need to accumulate memory should think hard about how to do it.

SparkQA commented:
**[Test build #51010 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51010/consoleFull)** for PR 11141 at commit [`7b54584`](https://github.com/apache/spark/commit/7b545841d50f08f40da8b52845f1dadab1731606).

SparkQA commented:
**[Test build #51010 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51010/consoleFull)** for PR 11141 at commit [`7b54584`](https://github.com/apache/spark/commit/7b545841d50f08f40da8b52845f1dadab1731606).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51010/
Test PASSed.

nongli committed:
[SPARK-13250][SQL] Update PhysicallRDD to convert to UnsafeRow if using the vectorized scanner.

Some parts of the engine rely on UnsafeRow which the vectorized parquet scanner does not want
to produce. This add a conversion in Physical RDD. In the case where codegen is used (and the
scan is the start of the pipeline), there is no requirement to use UnsafeRow. This patch adds
update PhysicallRDD to support codegen, which eliminates the need for the UnsafeRow conversion
in all cases.

The result of these changes for TPCDS-Q19 at the 10gb sf reduces the query time from 9.5 seconds
to 6.5 seconds.

nongli commented:
I updated the approach to do this a differnet way.

SparkQA commented:
**[Test build #51070 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51070/consoleFull)** for PR 11141 at commit [`4b6093a`](https://github.com/apache/spark/commit/4b6093a2788e693318debcdf4a4194667598d450).

SparkQA commented:
**[Test build #51070 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51070/consoleFull)** for PR 11141 at commit [`4b6093a`](https://github.com/apache/spark/commit/4b6093a2788e693318debcdf4a4194667598d450).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51070/
Test FAILed.

davies commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala:
```diff
@@ -118,6 +121,32 @@ private[sql] case class PhysicalRDD(
+
+  // Support codegen so that we can avoid the UnsafeRow conversion in all cases. Codegen
+  // never requires UnsafeRow as input.
+  override def supportCodegen: Boolean = true
```
It's true by default in CodegenSupport

davies commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala:
```diff
@@ -101,10 +102,12 @@ private[sql] case class PhysicalRDD(
-    if (isUnsafeRow) {
+    val conf = SQLContext.getActive().get
+    // The vectorized reader does not produce UnsafeRows. In this case we will convert.
+    if (isUnsafeRow && !conf.getConf(SQLConf.PARQUET_VECTORIZED_READER_ENABLED)) {
```
We should also check that the data source is Parquet or not. I'd suggest to do this in `createFromDataSource`

davies commented:
Could you also update the title and description of PR?

nongli committed:
CR

SparkQA commented:
**[Test build #51071 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51071/consoleFull)** for PR 11141 at commit [`a5448a3`](https://github.com/apache/spark/commit/a5448a34756bea26dec6ebb8ac2c56efc16185f1).

SparkQA commented:
**[Test build #51071 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51071/consoleFull)** for PR 11141 at commit [`a5448a3`](https://github.com/apache/spark/commit/a5448a34756bea26dec6ebb8ac2c56efc16185f1).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51071/
Test FAILed.

nongli committed:
Fix import ordering.

SparkQA commented:
**[Test build #51073 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51073/consoleFull)** for PR 11141 at commit [`b18ab65`](https://github.com/apache/spark/commit/b18ab65f2ce3715f7550a8fdfd4a0e3462ca2584).

SparkQA commented:
**[Test build #51073 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51073/consoleFull)** for PR 11141 at commit [`b18ab65`](https://github.com/apache/spark/commit/b18ab65f2ce3715f7550a8fdfd4a0e3462ca2584).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51073/
Test FAILed.

nongli committed:
Fix tests.

SparkQA commented:
**[Test build #51081 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51081/consoleFull)** for PR 11141 at commit [`bad9f33`](https://github.com/apache/spark/commit/bad9f33ff3ee1018eae3ec36339f051a28562fa6).

SparkQA commented:
**[Test build #51081 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51081/consoleFull)** for PR 11141 at commit [`bad9f33`](https://github.com/apache/spark/commit/bad9f33ff3ee1018eae3ec36339f051a28562fa6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51081/
Test FAILed.

nongli committed:
Fix a few tests that check query plan to disable whole stage codegen.

SparkQA commented:
**[Test build #51203 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51203/consoleFull)** for PR 11141 at commit [`3dbfc52`](https://github.com/apache/spark/commit/3dbfc524c637091b709b41bbfc37a1a705a7120b).

SparkQA commented:
**[Test build #51203 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51203/consoleFull)** for PR 11141 at commit [`3dbfc52`](https://github.com/apache/spark/commit/3dbfc524c637091b709b41bbfc37a1a705a7120b).
 * This patch **fails PySpark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51203/
Test FAILed.

davies commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala:
```diff
@@ -118,6 +120,30 @@ private[sql] case class PhysicalRDD(
+       | while (input.hasNext()) {
+       |   InternalRow $row = (InternalRow) input.next();
+       |   ${columns.map(_.code).mkString("\n").trim}
+       |   ${consume(ctx, columns).trim}
```
Could you add SQL metric for "numOutputRows"?

nongli committed:
Fix python plan test.

nongli committed:
Merge fixes

SparkQA commented:
**[Test build #51788 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51788/consoleFull)** for PR 11141 at commit [`534c472`](https://github.com/apache/spark/commit/534c4728b12a0143c8a76ba5352204eba3f2c427).

SparkQA commented:
**[Test build #2573 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2573/consoleFull)** for PR 11141 at commit [`534c472`](https://github.com/apache/spark/commit/534c4728b12a0143c8a76ba5352204eba3f2c427).

SparkQA commented:
**[Test build #51788 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51788/consoleFull)** for PR 11141 at commit [`534c472`](https://github.com/apache/spark/commit/534c4728b12a0143c8a76ba5352204eba3f2c427).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51788/
Test FAILed.

nongli committed:
Update test.

SparkQA commented:
**[Test build #51823 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51823/consoleFull)** for PR 11141 at commit [`19ebdbb`](https://github.com/apache/spark/commit/19ebdbb4fa24f8d0874cfeaa4004fa265c228e87).

SparkQA commented:
**[Test build #51823 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51823/consoleFull)** for PR 11141 at commit [`19ebdbb`](https://github.com/apache/spark/commit/19ebdbb4fa24f8d0874cfeaa4004fa265c228e87).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51823/
Test PASSed.

davies commented:
LGTM, @nongli Could you fix the conflicts?

nongli committed:
Merge fix

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51885/
Test FAILed.

SparkQA commented:
**[Test build #51887 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51887/consoleFull)** for PR 11141 at commit [`955e9e4`](https://github.com/apache/spark/commit/955e9e4673848d4783e46eb594d3f3f3d9fbc8cc).

nongli committed:
Merge conflicts.

SparkQA commented:
**[Test build #51887 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51887/consoleFull)** for PR 11141 at commit [`955e9e4`](https://github.com/apache/spark/commit/955e9e4673848d4783e46eb594d3f3f3d9fbc8cc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51887/
Test FAILed.

yhuai commented:
test this please

SparkQA commented:
**[Test build #51893 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51893/consoleFull)** for PR 11141 at commit [`731c30e`](https://github.com/apache/spark/commit/731c30e1d6653f050d480f7fd8be60e54cd6ffff).

SparkQA commented:
**[Test build #51893 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51893/consoleFull)** for PR 11141 at commit [`731c30e`](https://github.com/apache/spark/commit/731c30e1d6653f050d480f7fd8be60e54cd6ffff).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51893/
Test FAILed.

yhuai commented:
test this please

SparkQA commented:
**[Test build #51908 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51908/consoleFull)** for PR 11141 at commit [`731c30e`](https://github.com/apache/spark/commit/731c30e1d6653f050d480f7fd8be60e54cd6ffff).

SparkQA commented:
**[Test build #51908 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51908/consoleFull)** for PR 11141 at commit [`731c30e`](https://github.com/apache/spark/commit/731c30e1d6653f050d480f7fd8be60e54cd6ffff).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51908/
Test PASSed.

davies commented:
Merging this into master, thanks!
